### PR TITLE
test(dm): fix unsteable test TestFailToStartLeader

### LINKED ISSUE #4565: DM TestFailToStartLeader
### Which jobs are flaking?

https://ci.pingcap.net/blue/organizations/jenkins/ut-check/detail/ut-check/1237/pipeline/

### Which test(s) are flaking?

TestFailToStartLeader

### Jenkins logs or GitHub Actions link

```log
[2022-02-10T10:59:54.496Z] goroutine 300 [semacquire, 4 minutes]:
[2022-02-10T10:59:54.496Z] sync.runtime_Semacquire(0xc001b6e090)
[2022-02-10T10:59:54.496Z] 	/usr/local/go/src/runtime/sema.go:56 +0x45
[2022-02-10T10:59:54.496Z] sync.(*WaitGroup).Wait(0xc001b6e088)
[2022-02-10T10:59:54.496Z] 	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
[2022-02-10T10:59:54.496Z] github.com/pingcap/tiflow/dm/dm/master.(*Server).Close(0xc001b6e000)
[2022-02-10T10:59:54.496Z] 	/home/jenkins/agent/workspace/ut-check/tiflow/dm/dm/master/server.go:275 +0x15d
[2022-02-10T10:59:54.496Z] runtime.Goexit()
[2022-02-10T10:59:54.496Z] 	/usr/local/go/src/runtime/panic.go:632 +0x146
[2022-02-10T10:59:54.496Z] github.com/pingcap/check.(*C).stopNow(...)
[2022-02-10T10:59:54.496Z] 	/go/pkg/mod/github.com/pingcap/check@v0.0.0-20211026125417-57bd13f7b5f0/check.go:108
[2022-02-10T10:59:54.496Z] github.com/pingcap/check.(*C).Assert(0xc0002b2c30, 0x5a8b180, 0x92a7060, 0x71e7438, 0x8d882c0, 0xc0010d59d8, 0x1, 0x1)
[2022-02-10T10:59:54.496Z] 	/go/pkg/mod/github.com/pingcap/check@v0.0.0-20211026125417-57bd13f7b5f0/helpers.go:179 +0xc5
[2022-02-10T10:59:54.497Z] github.com/pingcap/tiflow/dm/dm/master.(*testElectionSuite).TestFailToStartLeader(0x9f17d98, 0xc0002b2c30)
[2022-02-10T10:59:54.497Z] 	/home/jenkins/agent/workspace/ut-check/tiflow/dm/dm/master/election_test.go:79 +0x103a
[2022-02-10T10:59:54.497Z] reflect.Value.call(0x5b4b8a0, 0x9f17d98, 0x213, 0x609a1a2, 0x4, 0xc0000e3ef0, 0x1, 0x1, 0xc000583d08, 0x4b4b4de, ...)
[2022-02-10T10:59:54.497Z] 	/usr/local/go/src/reflect/value.go:476 +0x99c
[2022-02-10T10:59:54.497Z] reflect.Value.Call(0x5b4b8a0, 0x9f17d98, 0x213, 0xc000583ef0, 0x1, 0x1, 0x0, 0x0, 0x0)
[2022-02-10T10:59:54.497Z] 	/usr/local/go/src/reflect/value.go:337 +0xd9
[2022-02-10T10:59:54.497Z] github.com/pingcap/check.(*suiteRunner).forkTest.func1(0xc0002b2c30)
[2022-02-10T10:59:54.497Z] 	/go/pkg/mod/github.com/pingcap/check@v0.0.0-20211026125417-57bd13f7b5f0/check.go:850 +0xb3c
[2022-02-10T10:59:54.497Z] github.com/pingcap/check.(*suiteRunner).forkCall.func1(0xc001149830, 0xc0002b2c30, 0xc000bd2ea0)
[2022-02-10T10:59:54.497Z] 	/go/pkg/mod/github.com/pingcap/check@v0.0.0-20211026125417-57bd13f7b5f0/check.go:739 +0x11e
[2022-02-10T10:59:54.497Z] created by github.com/pingcap/check.(*suiteRunner).forkCall
[2022-02-10T10:59:54.497Z] 	/go/pkg/mod/github.com/pingcap/check@v0.0.0-20211026125417-57bd13f7b5f0/check.go:734 +0x4cd
[2022-02-10T10:59:54.497Z] 
```

https://ci.pingcap.net/blue/organizations/jenkins/ut-check/detail/ut-check/1237/pipeline/

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve?

Issue Number: close #4565

### What is changed and how it works?

wait until master ready

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### Release note 

```release-note
Please add a release note.
`None`.
```

### CODE REVIEW

File: dm/dm/master/election_test.go [hunk 1]
```diff
@@ -33,7 +33,6 @@ type testElectionSuite struct{}
 
 func (t *testElectionSuite) TestFailToStartLeader(c *check.C) {
 	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
```
Comment (lance6716) [inline dm/dm/master/election_test.go:36]: I think it's OK to keep this. If c.Asset in line 40 failed, there will be a temporary context leak (although no harm)

File: dm/dm/master/election_test.go [hunk 2]
```diff
@@ -67,6 +66,11 @@ func (t *testElectionSuite) TestFailToStartLeader(c *check.C) {
 	s2 := NewServer(cfg2)
 	c.Assert(s2.Start(ctx), check.IsNil)
 	defer s2.Close()
+	defer cancel() // this cancel must call before s.Close() to avoid deadlock
```
Comment (lance6716) [inline dm/dm/master/election_test.go:69]: what about `defer s1.Close()` in line 49?

**Review by lance6716 — APPROVED** [review]

**Comment by Ehco1996** [comment]
could you please take another look for this pr ? thanks 🧡🧡🧡

/cc @D3Hunter

**Review by D3Hunter — APPROVED** [review]

File: dm/dm/master/election_test.go [hunk 3]
```diff
@@ -67,6 +68,12 @@ func (t *testElectionSuite) TestFailToStartLeader(c *check.C) {
...
```
Comment (GMHDBJD) [inline dm/dm/master/election_test.go:71]: why defer cancel three times? L36,L51,L71,L114

Comment (Ehco1996) [reply dm/dm/master/election_test.go:71]: see this https://github.com/pingcap/tiflow/pull/4556#discussion_r804429264

Comment (GMHDBJD) [reply dm/dm/master/election_test.go:71]: How about defer func(){
    cancel()
    s1.close()
    s2.close()
}()
What you need is to pre declare s1 and s2

Comment (Ehco1996) [reply dm/dm/master/election_test.go]: added in https://github.com/pingcap/tiflow/pull/4566/commits/7c6509b92cd1895c9beec2040e41885ab2812a04

File: dm/dm/master/election_test.go [hunk 4]
```diff
@@ -33,7 +33,13 @@ type testElectionSuite struct{}
 
 func (t *testElectionSuite) TestFailToStartLeader(c *check.C) {
 	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+
+	var s1, s2 *Server
+	defer func() {
+		cancel()
+		s1.Close()
+		s2.Close()
```
Comment (D3Hunter) [inline dm/dm/master/election_test.go:41]: s2 maybe nil and panic

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**